pallets/crowdloans/src/lib.rs

```diff
@@ -959,6 +976,154 @@ pub mod pallet {
+                    all_contributed_refunded = true;
+                    return Ok(());
+                }
+                contributed_refunded_count = contributed_refunded_count + 1;
+
+                // Mint relay currency
+                T::Assets::mint_into(T::RelayCurrency::get(), &who, amount)?;
+                Self::contribution_kill(
+                    T::RemoveKeysLimit::get(), &who, ChildStorageKind::Contributed
+                );
+            }
+
+            // If all contributions have been refunded then return AllRefunded event
+            // if not then return PartiallyRefunded event
+            if all_pending_refunded && all_contributed_refunded {
+                Self::deposit_event(Event::<T>::AllRefunded(
+                    crowdloan
+                ));
+            } else {
+                Self::deposit_event(Event::<T>::PartiallyRefunded(
+                    crowdloan,
+                ));
+            }
+
+            Ok(())
+        }
+
+        #[pallet::weight(<T as Config>::WeightInfo::dissolve_vault())]
+        #[transactional]
+        pub fn dissolve_vault(
+            origin: OriginFor<T>,
+            crowdloan: ParaId,
+            lease_start: LeasePeriod,
+            lease_end: LeasePeriod,
+        ) -> DispatchResult {
+            // users who can create vaults can dissolve them
+            T::DissolveVaultOrigin::ensure_origin(origin)?;
+
+            // 1. check phase, should be Closed or Failed or Expired
+            let vault = Self::vaults((&crowdloan, &lease_start, &lease_end))
+                .ok_or(Error::<T>::VaultDoesNotExist)?;
+
+            ensure!(
+                vault.phase == VaultPhase::Closed
+                    || vault.phase == VaultPhase::Failed
+                    || vault.phase == VaultPhase::Expired,
+                Error::<T>::IncorrectVaultPhase
+            );
+
+            // 2. check flying, pending, contributed childstorage, should have 0 contributions and `vault.contributed + vault.flying + vault.pending == 0` otherwise return back `NotReadyToDissolve`
+            let total_completed_contributions =
+                Self::contribution_iterator(vault.trie_index, ChildStorageKind::Contributed)
+                    .fold(0u128, |sum, (_account, (amount, _ref_code))| sum + amount);
+
+            let total_pending_contributions =
+                Self::contribution_iterator(vault.trie_index, ChildStorageKind::Pending)
+                    .fold(0u128, |sum, (_account, (amount, _ref_code))| sum + amount);
+
+            let total_flying_contributions =
+                Self::contribution_iterator(vault.trie_index, ChildStorageKind::Flying)
+                    .fold(0u128, |sum, (_account, (amount, _ref_code))| sum + amount);
+            
+            // TODO: fix this 
+            ensure!(
+                total_completed_contributions
+                .checked_add(total_flying_contributions)
```

Comment:
we have a function called `total_contribution`, you can reuse it by removing its `amount` argument